src/Microsoft.ML.Core/Properties/AssemblyInfo.cs

```diff
@@ -12,6 +12,7 @@
 [assembly: InternalsVisibleTo(assemblyName: "Microsoft.ML.InferenceTesting" + PublicKey.TestValue)]
 [assembly: InternalsVisibleTo(assemblyName: "Microsoft.ML.StaticPipelineTesting" + PublicKey.TestValue)]
 [assembly: InternalsVisibleTo(assemblyName: "Microsoft.ML.OnnxTransformTest" + PublicKey.TestValue)]
+[assembly: InternalsVisibleTo(assemblyName: "Microsoft.ML.BenchMarks" + PublicKey.TestValue)]
```

Comment:
Surely this isn't right, is it? Our Benchmarks should be written from an end-user's perspective, and not depend on internals.

Reply:
I'll remove it since I think it helps improve the benchmark code to start using the more modern APIs, but I'm not sure you're right about this? Why could we not use the benchmark project to test individual components?

I'll give a specific example. In some other work currently in progress, I have written benchmarks against FastTree's `IntArray`, as I'm exploring the possibility that the sumup routine can be purely managed. Currently that is a public class, but assuredly it must be internal in the future. Will I be disallowed from being able to use benchmarks for this? If we want to have two separate benchmark projects -- one for user facing stuff, one for our internal components -- that might be fine, but having no story is not one I consider an acceptable outcome.

---
In reply to: [234246392](https://github.com/dotnet/machinelearning/pull/1626#discussion_r234246392) [](ancestors = 234246392)

Reply:
Honestly, my opinion is that both functional and performance tests should work that way.

* Unit Tests 
  - they can have IVT the product, because they test units in isolation. and the best way to do that is to get at internal things.
* End-to-End Customer Tests
  - these should not have IVT the product, and should be more like what a customer would do/see.

Reply:
But a benchmark is not exclusively and end-to-end customer test. And benchmarking must not be only end-to-end stuff. We don't make a product with 1000 moving parts efficient by having end-to-end tests, because the reality is that any improvement we measure on *one* piece will be dwarfed by the other 999 pieces that were not improved. It is nonetheless crucial that we be aware of the perf of these small pieces, because only by doing this multiple times will we begin to see overall improvement in our strategy. Saying benchmarks may only be run on the highest level and not on internal components runs counter to this strategy.

Reply:
I might not have been clear in what I was saying above. I believe there are 4 types of tests:

* Perf end-to-end
* Perf unit
* Functional end-to-end
* Functional unit

Test Type| Unit | End to End
|------------|------------ | -------------
Functional Tests | IVT Allowed | No IVT 
Performance Tests | IVT Allowed | No IVT

Reply:
Sure, that's fine, but what are you suggesting we do with the internal perf tests? Separate project?

Reply:
(Obviously this isn't something we can do in this PR. But it would be a nice "medium-term goal".)